AssistedInject with ContributesBinding causes NullPointerException with generateContributionProviders

### Description

The compiler seems to throw a `NullPointerException` if an `@AssistedInject`-annotated class is also annotated with `@ContributesBinding`. While this is already incorrect, the error should probably be from a diagnostic, not from the compiler:

```
Caused by: java.lang.NullPointerException
	at dev.zacsweers.metro.compiler.ir.transformers.AssistedFactoryTransformer.generateCompanionDeclarations(AssistedFactoryTransformer.kt:292)
	at dev.zacsweers.metro.compiler.ir.transformers.AssistedFactoryTransformer.getOrGenerateImplClass$compiler(AssistedFactoryTransformer.kt:169)
	at dev.zacsweers.metro.compiler.ir.transformers.AssistedFactoryTransformer.visitClass(AssistedFactoryTransformer.kt:91)
	at dev.zacsweers.metro.compiler.ir.transformers.CoreTransformers.visitClassInner(CoreTransformers.kt:135)
	at dev.zacsweers.metro.compiler.ir.transformers.CoreTransformers.visitClassNew(CoreTransformers.kt:100)
	at org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext.visitClass(IrElementTransformerVoidWithContext.kt:62)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitClass(IrElementTransformerVoid.kt:57)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitClass(IrElementTransformerVoid.kt:19)
	at org.jetbrains.kotlin.ir.declarations.IrClass.accept(IrClass.kt:72)
	at org.jetbrains.kotlin.ir.IrElementBase.transform(IrElementBase.kt:33)
	at org.jetbrains.kotlin.ir.util.TransformKt.transformInPlace(transform.kt:35)
	at org.jetbrains.kotlin.ir.declarations.IrClass.transformChildren(IrClass.kt:82)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitDeclaration(IrElementTransformerVoid.kt:40)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitClass(IrElementTransformerVoid.kt:54)
	at org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext.visitClassNew(IrElementTransformerVoidWithContext.kt:126)
	at dev.zacsweers.metro.compiler.ir.transformers.CoreTransformers.visitClassNew(CoreTransformers.kt:101)
	at org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext.visitClass(IrElementTransformerVoidWithContext.kt:62)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitClass(IrElementTransformerVoid.kt:57)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitClass(IrElementTransformerVoid.kt:19)
	at org.jetbrains.kotlin.ir.declarations.IrClass.accept(IrClass.kt:72)
	at org.jetbrains.kotlin.ir.IrElementBase.transform(IrElementBase.kt:33)
	at org.jetbrains.kotlin.ir.util.TransformKt.transformInPlace(transform.kt:35)
	at org.jetbrains.kotlin.ir.declarations.IrPackageFragment.transformChildren(IrPackageFragment.kt:31)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitPackageFragment(IrElementTransformerVoid.kt:146)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitFile(IrElementTransformerVoid.kt:160)
	at org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext.visitFileNew(IrElementTransformerVoidWithContext.kt:122)
	at org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext.visitFile(IrElementTransformerVoidWithContext.kt:55)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitFile(IrElementTransformerVoid.kt:163)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitFile(IrElementTransformerVoid.kt:19)
	at org.jetbrains.kotlin.ir.declarations.IrFile.accept(IrFile.kt:27)
	at org.jetbrains.kotlin.ir.declarations.IrFile.transform(IrFile.kt:30)
	at org.jetbrains.kotlin.ir.declarations.IrFile.transform(IrFile.kt:19)
	at org.jetbrains.kotlin.ir.util.TransformKt.transformInPlace(transform.kt:35)
	at org.jetbrains.kotlin.ir.declarations.IrModuleFragment.transformChildren(IrModuleFragment.kt:46)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitModuleFragment(IrElementTransformerVoid.kt:102)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitModuleFragment(IrElementTransformerVoid.kt:107)
	at org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid.visitModuleFragment(IrElementTransformerVoid.kt:19)
	at org.jetbrains.kotlin.ir.declarations.IrModuleFragment.accept(IrModuleFragment.kt:36)
	at org.jetbrains.kotlin.ir.declarations.IrModuleFragment.transform(IrModuleFragment.kt:39)
	at dev.zacsweers.metro.compiler.ir.MetroIrGenerationExtension.generateInner(MetroIrGenerationExtension.kt:140)
	at dev.zacsweers.metro.compiler.ir.MetroIrGenerationExtension.generate(MetroIrGenerationExtension.kt:71)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.applyIrGenerationExtensions(convertToIr.kt:578)
	... 26 more
```

### Self-contained Reproducer

```kotlin
// GENERATE_CONTRIBUTION_PROVIDERS

interface Blah

@ContributesBinding(AppScope::class)
class RealBlah @AssistedInject constructor(
    @Assisted private val assistedString: String
): Blah {
    @AssistedFactory
    fun interface Factory {
        fun create(assistedString: String): RealBlah
    }
}

@DependencyGraph(AppScope::class)
interface AppGraph {
    val factory: RealBlah.Factory
}

fun box(): String {
    val graph = createGraph<AppGraph>()
    assertNotNull(graph.factory.create("hello"))
    return "OK"
}
```

### Metro version

1.0.0

### Previous working version

N/A

### Kotlin version

2.3.21

### IDE version

_No response_

### Platform

_No response_

### Context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ Changelog
 ### New
 
 - **[FIR]** Add a diagnostic warning that `@Module.subcomponents` is ignored by Metro (when Dagger interop is enabled).
+- **[FIR]** Add a diagnostic error for annotating `@AssistedInject` types with `@Contributes*` annotations, as they are not compatible.
 
 ### Enhancements
 

--- a/compiler-tests/src/test/data/diagnostic/inject/assisted/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.diag.txt
+++ b/compiler-tests/src/test/data/diagnostic/inject/assisted/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.diag.txt
@@ -1,0 +1,7 @@
+/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.kt:8:1: error: `@ContributesBinding` is not applicable to `@AssistedInject`-annotated classes. Remove `@AssistedInject` or apply the contribution elsewhere.
+
+/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.kt:19:1: error: `@ContributesIntoSet` is not applicable to `@AssistedInject`-annotated classes. Remove `@AssistedInject` or apply the contribution elsewhere.
+
+/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.kt:30:1: error: `@ContributesIntoMap` is not applicable to `@AssistedInject`-annotated classes. Remove `@AssistedInject` or apply the contribution elsewhere.
+
+/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.kt:42:1: error: `@ContributesBinding` is not applicable to `@AssistedInject`-annotated classes. Remove `@AssistedInject` or apply the contribution elsewhere.

--- a/compiler-tests/src/test/data/diagnostic/inject/assisted/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.kt
+++ b/compiler-tests/src/test/data/diagnostic/inject/assisted/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.kt
@@ -1,0 +1,55 @@
+// RENDER_DIAGNOSTICS_FULL_TEXT
+// SUPPRESS_WARNINGS: SUGGEST_CLASS_INJECTION
+
+interface Blah
+
+// Invalid: @ContributesBinding on a class with an @AssistedInject constructor
+<!AGGREGATION_ERROR!>@ContributesBinding(AppScope::class)<!>
+class CtorAssistedBinding @AssistedInject constructor(
+  @Assisted private val assistedString: String,
+) : Blah {
+  @AssistedFactory
+  fun interface Factory {
+    fun create(assistedString: String): CtorAssistedBinding
+  }
+}
+
+// Invalid: @ContributesIntoSet on a class with an @AssistedInject constructor
+<!AGGREGATION_ERROR!>@ContributesIntoSet(AppScope::class)<!>
+class CtorAssistedIntoSet @AssistedInject constructor(
+  @Assisted private val assistedString: String,
+) : Blah {
+  @AssistedFactory
+  fun interface Factory {
+    fun create(assistedString: String): CtorAssistedIntoSet
+  }
+}
+
+// Invalid: @ContributesIntoMap on a class with an @AssistedInject constructor
+<!AGGREGATION_ERROR!>@ContributesIntoMap(AppScope::class)<!>
+@StringKey("key")
+class CtorAssistedIntoMap @AssistedInject constructor(
+  @Assisted private val assistedString: String,
+) : Blah {
+  @AssistedFactory
+  fun interface Factory {
+    fun create(assistedString: String): CtorAssistedIntoMap
+  }
+}
+
+// Invalid: @ContributesBinding on a class-level @AssistedInject class
+<!AGGREGATION_ERROR!>@ContributesBinding(AppScope::class)<!>
+@AssistedInject
+class ClassLevelAssistedBinding(
+  @Assisted private val assistedString: String,
+) : Blah {
+  @AssistedFactory
+  fun interface Factory {
+    fun create(assistedString: String): ClassLevelAssistedBinding
+  }
+}
+
+// Valid: @ContributesBinding on a plain @Inject class (no assisted injection)
+@ContributesBinding(AppScope::class)
+@Inject
+class PlainInjectBinding : Blah

--- a/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/DiagnosticTestGenerated.java
+++ b/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/DiagnosticTestGenerated.java
@@ -918,6 +918,12 @@ public class DiagnosticTestGenerated extends AbstractDiagnosticTest {
       }
 
       @Test
+      @TestMetadata("ContributesAnnotationsNotAllowedOnAssistedInjectClasses.kt")
+      public void testContributesAnnotationsNotAllowedOnAssistedInjectClasses() {
+        runTest("compiler-tests/src/test/data/diagnostic/inject/assisted/ContributesAnnotationsNotAllowedOnAssistedInjectClasses.kt");
+      }
+
+      @Test
       @TestMetadata("InjectingAssistedInjectClassesIsAnError.kt")
       public void testInjectingAssistedInjectClassesIsAnError() {
         runTest("compiler-tests/src/test/data/diagnostic/inject/assisted/InjectingAssistedInjectClassesIsAnError.kt");

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/checkers/AggregationChecker.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/checkers/AggregationChecker.kt
@@ -10,6 +10,7 @@ import dev.zacsweers.metro.compiler.fir.MetroFirAnnotation
 import dev.zacsweers.metro.compiler.fir.annotationsIn
 import dev.zacsweers.metro.compiler.fir.classIds
 import dev.zacsweers.metro.compiler.fir.diagnosticString
+import dev.zacsweers.metro.compiler.fir.findAssistedInjectConstructors
 import dev.zacsweers.metro.compiler.fir.findInjectLikeConstructors
 import dev.zacsweers.metro.compiler.fir.isAnnotatedWithAny
 import dev.zacsweers.metro.compiler.fir.isBindingContainer
@@ -286,6 +287,18 @@ internal object AggregationChecker : FirClassChecker(MppCheckerKind.Common) {
         annotation.source,
         MetroDiagnostics.AGGREGATION_ERROR,
         "`@$kind` is only applicable to constructor-injected classes, assisted factories, or objects. Ensure ${declaration.symbol.classId.asSingleFqName()} is injectable or a bindable object.",
+      )
+      return false
+    }
+
+    if (
+      !isAssistedFactory &&
+        declaration.symbol.findAssistedInjectConstructors(session, checkClass = true).isNotEmpty()
+    ) {
+      reporter.reportOn(
+        annotation.source,
+        MetroDiagnostics.AGGREGATION_ERROR,
+        "`@$kind` is not applicable to `@AssistedInject`-annotated classes. Remove `@AssistedInject` or apply the contribution elsewhere.",
       )
       return false
     }

